source/utf8/unchecked.h

```diff
@@ -276,62 +276,73 @@ namespace utf8
             }
           }; // class bidirectional_iterator
 
-        template <typename octet_iterator>
-        class input_iterator : public std::iterator<std::input_iterator_tag, uint32_t> {
-          private:
-          octet_iterator it;
-          uint32_t cp{};
-          void read() {
-              cp = utf8::unchecked::next_impl(it);
-          }
-    
+        class iterator_stream {
           public:
-          using iterator_category = std::input_iterator_tag;
+          using Iter = std::istream_iterator<char>;
+
+          using iterator_category = Iter::iterator_category;
           using value_type = uint32_t;
           using difference_type = ptrdiff_t;
           using pointer = uint32_t*;
           using reference = uint32_t&;
 
-          input_iterator () {}
-          explicit input_iterator (const octet_iterator& octet_it) : it(octet_it)
+          private:
+          Iter it;
+          Iter end_range{};
+          uint32_t cp{};
+          bool ok{};
+
+          void read() {
+             ok = it != end_range;
+             if(ok) {
+              cp = utf8::unchecked::next_impl(it);
+             }
+          }
+
+          public:
+          iterator_stream () {}
+          explicit iterator_stream (const Iter& octet_it) : it(octet_it)
           {
               read();
           }
-          octet_iterator base () const { return it; }
+          Iter base () const { return it; }
           uint32_t operator * () const
           {
               return cp;
           }
     
-          bool operator == (const input_iterator& rhs) const
+          bool operator == (const iterator_stream& rhs) const
           {
-              return it == rhs.it;
+              return ok == rhs.ok && (!ok || it == rhs.it);
```

Comment:
This seems hard to verify. How about: `ok == rhs.ok && it == rhs.it;` and using `next` in `read`? This way:
- `it` always points to next CP or end
- if `it` points to end there are 2 cases
  - It is the `end` --> `ok == false`
  - It is 1 before `end` --> `ok==true`

This avoids the need for this kinda complicated check and the `next_impl` function

Reply:
Actually I don't need `ok` as a private member. A local variable `ok` in function `read()` is enough. The test `it == rhs.it;` is enough.

Reply:
I agree because `it` points to 1 before the next octet/byte and the assumption here is that the sequence is valid, complete UTF8. IMO the `next/next_impl` is rather strange and hard to grasp, so having the `ok` member here would avoid it. But well... I guess this is faster.

Reply:
Done